crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -1007,6 +1025,75 @@ pub(crate) fn has_own_parentheses(
     }
 }
 
+/// Returns `true` if the expression can hug directly to enclosing parentheses.
+///
+/// For example, in preview style, given:
+/// ```python
+/// ([1, 2, 3,])
+/// ```
+///
+/// We want to format it as:
+/// ```python
+/// ([
+///     1,
+///     2,
+///     3,
+/// ])
+/// ```
+///
+/// As opposed to:
+/// ```python
+/// (
+///     [
+///         1,
+///         2,
+///         3,
+///     ]
+/// )
+/// ```
+pub(crate) fn is_expression_huggable(expr: &Expr) -> bool {
+    match expr {
+        Expr::List(_)
+        | Expr::Set(_)
+        | Expr::Dict(_)
+        | Expr::ListComp(_)
+        | Expr::SetComp(_)
+        | Expr::DictComp(_) => true,
+
+        Expr::Starred(ast::ExprStarred { value, .. }) => matches!(
+            value.as_ref(),
+            Expr::List(_)
+                | Expr::Set(_)
+                | Expr::Dict(_)
+                | Expr::ListComp(_)
+                | Expr::SetComp(_)
+                | Expr::DictComp(_)
+        ),
+
+        Expr::BoolOp(_)
+        | Expr::NamedExpr(_)
+        | Expr::BinOp(_)
+        | Expr::UnaryOp(_)
+        | Expr::Lambda(_)
+        | Expr::IfExp(_)
+        | Expr::GeneratorExp(_)
+        | Expr::Await(_)
+        | Expr::Yield(_)
+        | Expr::YieldFrom(_)
+        | Expr::Compare(_)
+        | Expr::Call(_)
+        | Expr::FormattedValue(_)
+        | Expr::FString(_)
+        | Expr::Constant(_)
+        | Expr::Attribute(_)
+        | Expr::Subscript(_)
+        | Expr::Name(_)
+        | Expr::Tuple(_)
```

Comment:
Interesting that tuples are excluded.

Reply:
They re-added them in a later PR (https://github.com/psf/black/pull/4012), so I'm changing it here too.